sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/ActionDescriptorFactory.scala

```diff
@@ -48,11 +53,91 @@ private[impl] object ActionDescriptorFactory extends ComponentDescriptorFactory
+      def firstToUpper(name: String) = {
+        name.charAt(0).toUpper + name.substring(1)
+      }
+      groupByTopic(subscriptions).collect {
+        case (topic, kMethods) if kMethods.size > 1 =>
+          val representative = kMethods.head
+          val rs: RestServiceMethod = representative.serviceMethod.asInstanceOf[RestServiceMethod]
+          KalixMethod(
+            RestServiceMethod(rs.javaMethod, Some("On" + firstToUpper(topic) + firstToUpper(rs.javaMethod.getName))))
+            .withKalixOptions(representative.methodOptions)
+      }.toSeq
+    }
+
+    def groupByTopic(methods: Seq[KalixMethod]): Map[String, Seq[KalixMethod]] = {
+      val withTopicIn = methods.filter(kalixMethod =>
+        kalixMethod.methodOptions.exists(option =>
+          option.hasEventing && option.getEventing.hasIn && option.getEventing.getIn.hasTopic))
+      //Assuming there is only one topic, therefore head is as good as any other
+      withTopicIn.groupBy(m => m.methodOptions.head.getEventing.getIn.getTopic)
+    }
+
+    val publicationTopicMethods = component.getMethods
+      .filter(hasTopicPublication)
+      .sorted // make sure we get the methods in deterministic order
+      .map { method =>
+        val publicatinOptions = eventingOutForTopic(method)
+        val kalixOptions =
+          kalix.MethodOptions.newBuilder().setEventing(publicatinOptions).build()
+
+        KalixMethod(RestServiceMethod(method))
+          .withKalixOptions(kalixOptions)
+      }
     val serviceName = nameGenerator.getName(component.getSimpleName)
+
+    def filterAndAddKalixOptions(to: Seq[KalixMethod], from: Seq[KalixMethod]): Seq[KalixMethod] = {
+      val common = to.flatMap(toAdd =>
+        from
+          .filter { addingFrom =>
+            addingFrom.serviceMethod.methodName.equals(toAdd.serviceMethod.methodName)
+          }
+          .map(addingFrom => toAdd.withKalixOptions(addingFrom.methodOptions)))
+      val notInCommon = to
+        .filter { toAdd =>
+          !from.exists { addingFrom =>
+            addingFrom.serviceMethod.methodName.equals(toAdd.serviceMethod.methodName)
+          }
+        }
+      common ++ notInCommon
+    }
+
+    def removeDuplicates(springMethods: Seq[KalixMethod], pubSubMethods: Seq[KalixMethod]): Seq[KalixMethod] = {
+      pubSubMethods.filterNot(p =>
+        springMethods.exists(s => p.serviceMethod.methodName.equals(s.serviceMethod.methodName)))
+    }
+
     ComponentDescriptor(
       nameGenerator,
       serviceName,
       component.getPackageName,
-      springAnnotatedMethods ++ subscriptionMethods)
+      filterAndAddKalixOptions(
+        springAnnotatedMethods,
+        publicationTopicMethods) ++ subscriptionValueEntityMethods ++ combineSubscriptionsSameTopic(
+        subscriptionTopicMethods) ++ uniqueTopicSubscriptions(subscriptionTopicMethods) ++ removeDuplicates(
+        springAnnotatedMethods,
+        publicationTopicMethods))
```

Comment:
This block could benefit from some formatting to make it easier to read.

Reply:
done